src/omp/exec_thom.f90

```diff
@@ -0,0 +1,175 @@
+    type(tdsops_t), intent(in) :: tdsops
+
+    integer :: i, j, k
+    integer :: nk
+    integer :: ctr
+
+    nk = size(u, dim=3)
+    associate(coeffs_s => tdsops%coeffs_s, coeffs_e => tdsops%coeffs_e, coeffs => tdsops%coeffs, &
+              n => tdsops%tds_n, &
+              thom_f => tdsops%thom_f, thom_s => tdsops%thom_s, thom_w => tdsops%thom_w)
+      !$omp parallel do private(i, j, ctr)
+      do k = 1, nk
+        !! Forward pass
+        ! XXX: Extract as PURE helper?
+
+        !$omp simd
+        do i = 1, SZ
+          du(i, 1, k) = sum(coeffs_s(5:9, 1) * u(i, 1:5, k))
+        end do
+        !$omp end simd
+        do j = 2, 4
+          !$omp simd
+          do i = 1, SZ
+            du(i, j, k) = sum(coeffs_s(6-j:9, j) * u(i, 1:4+j, k)) &
+              - du(i, j - 1, k) * thom_s(j)
+          end do
+          !$omp end simd
+        end do
+       
+        do j = 5, n - 4
+          !$omp simd
+          do i = 1, SZ
+            du(i, j, k) = sum(coeffs(1:9) * u(i, j-4:j+4, k)) &
+              - du(i, j - 1, k) * thom_s(j)
+          end do
+          !$omp end simd
+        end do
+       
+        ctr = 0
+        do j = n - 3, n
+          !$omp simd
+          do i = 1, SZ
+            du(i, j, k) = sum(coeffs_e(1:8-ctr, ctr + 1) * u(i, j-4:j+(3-ctr), k)) &
+              - du(i, j - 1, k) * thom_s(j)
+          end do
+          !$omp end simd
+          ctr = ctr + 1
+        end do
+
+        !! Backward pass
+        ! XXX: Extract as PURE helper?
+
+        !$omp simd
+        do i = 1, SZ
+          du(i, n, k) = du(i, n, k) * thom_w(n)
+        end do
+        !$omp end simd
+        do j = n - 1, 1, -1
+          !$omp simd
+          do i = 1, SZ
+            du(i, j, k) = (du(i, j, k) - thom_f(j) * du(i, j + 1, k)) * thom_w(j)
+          end do
+          !$omp end simd
+        end do
+      end do
+      !$omp end parallel do
+    end associate
+    
+  end subroutine der_univ_thom
+
+  subroutine der_univ_thom_per(du, u, tdsops)
+
+    real(dp), dimension(:, :, :), intent(out) :: du
+    real(dp), dimension(:, :, :), intent(in) :: u
+    type(tdsops_t), intent(in) :: tdsops
+
+    integer :: nk
+    integer :: i, j, k
+
+    real(dp), dimension(SZ) :: ss
+    integer, dimension(4) :: jm, jp
+
+    nk = size(u, dim=3)
+    associate(coeffs_s => tdsops%coeffs_s, coeffs_e => tdsops%coeffs_e, coeffs => tdsops%coeffs, &
+              n => tdsops%tds_n, &
+              alpha => tdsops%alpha, &
+              thom_f => tdsops%thom_f, thom_s => tdsops%thom_s, thom_w => tdsops%thom_w, &
+              thom_p => tdsops%thom_p)
+      !$omp parallel do private(i, j, jm, jp, ss)
+      do k = 1, nk
+        !! Forward pass
+        do j = 1, n
+          do i = 1, 4
+            jm(i) = modulo(j - (1 + i), n) + 1
+            jp(i) = modulo(j - (n + (i - 1)), n) + 1
+          end do
+
+          !$omp simd
+          do i = 1, SZ
+            du(i, j, k) = sum(coeffs(1:4) * u(i, jm(1:4), k))
```

Comment:
The mapping between coeffs 1 to 4 and jm(1:4) seems like mismatched to indices provided via `jm` array. I think `jm` should be reversed here like `(4:1:-1)`. Also very curious if such index accesses can be vectorised. I don't trust compilers and instinctively avoid these but completely fine if we can achieve performance we're aiming.